connect/src/main/java/com/stripe/android/connect/webview/StripeDownloadListener.kt

```diff
@@ -0,0 +1,97 @@
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.MainScope
+import kotlinx.coroutines.delay
+import kotlinx.coroutines.launch
+
+internal class StripeDownloadListener(
+    private val context: Context,
+    private val downloadManager: DownloadManager? =
+        context.getSystemService(Context.DOWNLOAD_SERVICE) as? DownloadManager,
+    private val ioScope: CoroutineScope = CoroutineScope(Dispatchers.IO),
+    // methods below exposed for mocking in tests
+    private val getDownloadManagerRequest: (Uri) -> DownloadManager.Request = { uri ->
+        DownloadManager.Request(uri)
+    },
+    private val getFileName: (url: String, contentDisposition: String?, mimeType: String?) -> String =
+        { url, contentDisposition, mimeType ->
+            URLUtil.guessFileName(url, contentDisposition, mimeType)
+        },
+    private val parseUri: (String) -> Uri = { Uri.parse(it) },
+    private val showToast: (String) -> Unit = { toastString ->
+        Toast.makeText(context, toastString, Toast.LENGTH_LONG).show()
+    }
+) : DownloadListener {
+
+    override fun onDownloadStart(
+        url: String?,
+        userAgent: String?,
+        contentDisposition: String?,
+        mimetype: String?,
+        contentLength: Long
+    ) {
+        if (url == null || downloadManager == null) {
+            showErrorToast()
+            return
+        }
+
+        ioScope.launch {
+            val request = getDownloadManagerRequest(parseUri(url))
+            val fileName = getFileName(url, contentDisposition, mimetype)
+            request.setDestinationInExternalPublicDir(DIRECTORY_DOWNLOADS, fileName)
+            request.setNotificationVisibility(VISIBILITY_VISIBLE_NOTIFY_COMPLETED)
+            request.setTitle(fileName)
+            request.setDescription(context.getString(R.string.stripe_downloading_file))
+            request.setMimeType(mimetype)
+
+            val downloadId = downloadManager.enqueue(request)
+
+            // Monitor the download progress and show a toast when done
+            val query = DownloadManager.Query().setFilterById(downloadId)
+            var isDownloading = true
+            while (isDownloading) {
+                downloadManager.query(query).use { cursor ->
+                    cursor.moveToFirst()
+                    val index = cursor.getColumnIndex(DownloadManager.COLUMN_STATUS)
+                    if (index < 0) return@launch // status does not exist - abort
+                    val status = cursor.getInt(index)
+                    if (status !in listOf(STATUS_PENDING, STATUS_RUNNING, STATUS_PAUSED)) {
+                        showOpenFileToast()
+                        isDownloading = false // download complete - exit the loop
+                    }
+                }
+                delay(DOWNLOAD_DELAY_MS)
+            }
+        }
+    }
+
+    private fun showErrorToast() {
+        MainScope().launch {
```

Comment:
i'd pass in the main scope like you're doing with io Scope to make it more testable